internal/container/layer.go

```diff
@@ -0,0 +1,254 @@
+		if len(missingLayers) == 0 {
+			ret := files
+			files = nil // prevents the defer from closing the files
+			return ret, nil
+		}
+
+		// Some of the layer files were missing, we need to fetch and extract
+		// them from the object store. On success, the returned list of open
+		// files is guaranteed to have as many files as there were missing
+		// layers, with a 1:1 correspondence on the indexes.
+		missingFiles, err := ExtractLayers(ctx, dst, src, missingLayers)
+		if err != nil {
+			return nil, err
+		}
+
+		j := 0
+		for i, f := range files {
+			if f == nil {
+				files[i] = missingFiles[j]
+				j++
+			}
+		}
+	}
+}
+
+// ExtractLayers fetches the image layers represented by the list of descriptors
+// and extracts them at the dst file path.
+//
+// The content of the layers is written in a directory named after their digest
+// algorithm, and uncompressed in a tar archive named with their digest hash.
+//
+//	(dst)
+//	 └── sha256
+//	    └── 8c6d1654570f041603f4cef49c320c8f6f3e401324913009d92a19132cbf1ac0
+//	        ...
+//
+// The function performs the extraction of each layer concurrently, each layer
+// is processed in a dedicated goroutine.
+//
+// The operation is atomic, either the creation of the layer fully succeeds or
+// it fails and the layer is not present in the directory.
+func ExtractLayers(ctx context.Context, dst string, src oras.ReadOnlyTarget, layers []ocispec.Descriptor) ([]*os.File, error) {
+	files := make([]*os.File, len(layers))
+	errch := make(chan error)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for i, desc := range layers {
+		go func(i int, desc ocispec.Descriptor) {
+			f, err := ExtractLayer(ctx, dst, src, desc)
+			if err != nil {
+				cancel()
+				err = fmt.Errorf("%s: extract: %w", desc.Digest, err)
+			} else {
+				files[i] = f
+			}
+			errch <- err
+		}(i, desc)
+	}
+
+	var errs []error
+	for range layers {
+		if err := <-errch; err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := errors.Join(errs...); err != nil {
+		for _, f := range files {
+			if f != nil {
+				f.Close()
+			}
+		}
+		return nil, err
+	}
+
+	return files, nil
+}
+
+// ExtractLayer is like ExtractLayers but applied to a single layer.
+func ExtractLayer(ctx context.Context, dst string, src oras.ReadOnlyTarget, desc ocispec.Descriptor) (*os.File, error) {
+	algo := desc.Digest.Algorithm()
+	hash := desc.Digest.Encoded()
+	path := filepath.Join(dst, algo.String())
+
+	if err := os.Mkdir(path, 0777); err != nil {
+		if !errors.Is(err, fs.ErrExist) {
+			return nil, err
+		}
+	}
+
+	f, err := os.CreateTemp(path, "."+hash+".*")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// This boolean controls whether we need to remove the temporary file, it is
+	// set to true after the archive was extracted and moved to its final path,
+	// in which case the temporary file path does not exist anymore.
+	success := false
+	defer func() {
+		if !success {
+			os.Remove(f.Name())
+		}
+	}()
```

Comment:
Nit: `success == (err == nil)`; no need for an extra var and explanation? I think using `err` is technically more correct too, since `duplicateFile` could fail after `success = true` is set?